Add code of other CUDA packages

maleadt opened the pull request:
This PR merges in the code from CUDAapi.jl, CUDAdrv.jl, CUDAnative.jl and CuArrays.jl. There's been more and more coupling between these packages, and user confusion about what is provided by each of these packages, that it makes sense to try this.

The PR is mostly complete: Adds all functionality, minus the wrapper scripts. I'll merge these when the time comes to regenerate the wrappers. Documentation is TODO as well, but at least we'll now be able to host all docs in a single place.

I did the merge using `git subtree`, bringing in all history. A single commit then put all files in their proper location, that seemed much easier than try and merge them into place.

Finally, this PR also adds some functionality: parallel testing, since this package is now big enough to warrant that. Also adds `CUDA.@timed` returning GC statistics, used by the testing infrastructure.

leios commented:
Also: this doesn't necessarily need to be pulled into the main branch, I just figured I would be explicit here.

janEbert committed:
Implement nnz, nonzeros and nonzeroinds

This also fixes `show` for sparse GPU vectors.

janEbert committed:
Test nnz, nonzeros and nonzeroinds

thomasfaingnaert committed:
Fix return type of WMMA intrinsics

maleadt committed:
Destroy handles when the task finalizes.

Works around atext thread-safety issues.

wongalvis14 committed:
Return to a two-staged kernel, but retain the loop to process multiple elements per thread.

maleadt committed:
Allow the split allocator to fail, for now.

maleadt committed:
Merge pull request #645 from JuliaGPU/tb/multitasking

Support for Julia's multitasking.

maleadt committed:
Require more recent, released CUDAdrv.

maleadt committed:
Bump version. [ci skip]

maleadt committed:
Don't use deprecated env var.

maleadt committed:
Add missing dep for testing Flux.

maleadt committed:
Bump version. [ci skip]

maleadt committed:
Fix and test initialization without CUDA.

maleadt committed:
Test initialization without CUDA.

maleadt committed:
Fix and test initialization without CUDA.

maleadt committed:
Merge pull request #228 from JuliaGPU/tb/non_functional

Fix and test initialization without CUDA.

maleadt committed:
Bump version. [ci skip]

maleadt committed:
Merge pull request #650 from JuliaGPU/tb/non_functional

Fix and test initialization without CUDA.

maleadt committed:
Bump version. [ci skip]

maleadt committed:
Improved error output during finalizer exceptions.

maleadt committed:
Bump CI to include 1.4.

maleadt committed:
Disable WMMA tests when running a debug build.

maleadt committed:
Merge pull request #616 from JuliaGPU/tb/ci_1.4

Bump CI to 1.4, and include debug build

maleadt committed:
Merge pull request #615 from JuliaGPU/tb/non_functional

Better test initialization without CUDA.

maleadt committed:
Merge pull request #651 from JuliaGPU/tb/finalizer_errors

Improved error output during finalizer exceptions.

maleadt committed:
Adapt to restructured codegen.

maleadt committed:
Fix WMMA codegen tests for below PTX 6.3.

maleadt committed:
Merge pull request #162 from JuliaGPU/tb/codegen-norecursion

Use new non-recursive codegen

maleadt committed:
Only run WMMA tests on a compatible debug build.

maleadt committed:
Merge pull request #591 from thomasfaingnaert/tf/fix-wmma-rettype

Fix return type of WMMA intrinsics

maleadt committed:
Silence a message. [ci skip]

maleadt committed:
Bump version. [ci skip]

maleadt committed:
Switch to ExprTools.

maleadt committed:
Merge pull request #617 from JuliaGPU/tb/exprtools

Switch to ExprTools.

maleadt committed:
Don't double-count freed memory.

maleadt committed:
Prevent potential deadlocks by not taking multiple locks at the same time.

maleadt committed:
Deduplicate irgen.

maleadt committed:
Merge pull request #654 from JuliaGPU/tb/usage_discrepancy

Fix usage discrepancies

maleadt committed:
Merge pull request #618 from JuliaGPU/tb/deduplicate

Deduplicate irgen.

chrstphrbrns committed:
add support for uniform random UInt32

github-actions[bot] committed:
Update dependencies.

Project CUDAnative v3.0.2
    Status `~/work/CUDAnative.jl/CUDAnative.jl/Project.toml`
  [79e6a3ab]   Adapt v1.0.1
  [b99e7846]   BinaryProvider v0.5.8
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814] ↑ CUDAdrv v6.2.1 ⇒ v6.2.2
  [f68482b8]   Cthulhu v1.0.0
  [864edb3b]   DataStructures v0.17.10
  [e2ba6199]   ExprTools v0.1.0
  [929cbde3]   LLVM v1.3.4
  [a759f4b9]   TimerOutputs v0.5.3
  [b77e0a4c]   InteractiveUtils
  [8f399da3]   Libdl
  [44cfe95a]   Pkg
  [de0858da]   Printf

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.0.1
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814] ↑ CUDAdrv v6.2.0 ⇒ v6.2.2
  [be33ccc6] ↑ CUDAnative v3.0.0 ⇒ v3.0.2
  [864edb3b]   DataStructures v0.17.10
  [0c68f7d7]   GPUArrays v3.1.0
  [1914dd2f] ↑ MacroTools v0.5.4 ⇒ v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9]   TimerOutputs v0.5.3
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #659 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Merge pull request #619 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Merge pull request #656 from chrstphrbrns/rand-uint32

Add support for uniform random UInt32

maleadt committed:
Implement GPUArrays vararg mapreduce.

maleadt committed:
Merge pull request #663 from JuliaGPU/tb/mapreduce_vararg

Implement GPUArrays vararg mapreduce.

maleadt committed:
Figure out the block sizes without passing additional arguments.

maleadt committed:
Better align the launch configuration with what the occupancy API returns.

bors[bot] committed:
Merge #646

646: Improve mapreduce performance r=maleadt a=wongalvis14

~More than 3-fold improvement over the latest implementation~

Benchmarking function from https://github.com/JuliaGPU/CuArrays.jl/issues/611

First stage: Using the number of "max parallel threads a single block can hold" as the number of blocks, perform reduction with serial iteration if needed

Second stage: Reduction in a single block, no serial iteration

This approach aims to strike an optimal balance between workload of each thread, kernel launch overhead and parallel resource exhaustion.

```
New impl:
julia> @benchmark pi_mc_cu(10000000)
BenchmarkTools.Trial: 
  memory estimate:  16.98 KiB
  allocs estimate:  468
  --------------
  minimum time:     2.520 ms (0.00% GC)
  median time:      2.536 ms (0.00% GC)
  mean time:        2.584 ms (0.64% GC)
  maximum time:     15.600 ms (50.62% GC)
  --------------
  samples:          1930
  evals/sample:     1

Old recursion impl:
julia> @benchmark pi_mc_cu(10000000)
BenchmarkTools.Trial: 
  memory estimate:  17.05 KiB
  allocs estimate:  472
  --------------
  minimum time:     4.059 ms (0.00% GC)
  median time:      4.076 ms (0.00% GC)
  mean time:        4.130 ms (0.64% GC)
  maximum time:     23.199 ms (63.12% GC)
  --------------
  samples:          1209
  evals/sample:     1

Latest serial impl:
BenchmarkTools.Trial: 
  memory estimate:  7.81 KiB
  allocs estimate:  242
  --------------
  minimum time:     8.544 ms (0.00% GC)
  median time:      8.579 ms (0.00% GC)
  mean time:        8.622 ms (0.27% GC)
  maximum time:     26.172 ms (41.80% GC)
  --------------
  samples:          580
  evals/sample:     1
```

Co-authored-by: wongalvis14 <wongalvis14@gmail.com>
Co-authored-by: Tim Besard <tim.besard@gmail.com>

maleadt committed:
Test for implicit singleton dims with mapreducedim.

maleadt committed:
Merge pull request #665 from JuliaGPU/tb/mapreduce_implicit_singleton

Test for implicit singleton dims with mapreducedim.

maleadt committed:
Update README.md

janEbert committed:
Implement getindex

This also fixes `show` for sparse GPU matrices.

janEbert committed:
Test getindex

kshyatt committed:
Small missing tests for device printing and eltype

kshyatt committed:
Additional tests for stream equality

kshyatt committed:
Test for bad pointer conversion

kshyatt committed:
Test for prefetch error

kshyatt committed:
Return prefetch test to life

kshyatt committed:
Get rid of references to non-existent functions in docs, add info about flags

github-actions[bot] committed:
Update dependencies.

Project CUDAnative v3.0.2
    Status `~/work/CUDAnative.jl/CUDAnative.jl/Project.toml`
  [79e6a3ab]   Adapt v1.0.1
  [b99e7846]   BinaryProvider v0.5.8
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.2
  [f68482b8] ↑ Cthulhu v1.0.0 ⇒ v1.0.1
  [864edb3b] ↑ DataStructures v0.17.10 ⇒ v0.17.11
  [e2ba6199]   ExprTools v0.1.0
  [929cbde3]   LLVM v1.3.4
  [a759f4b9]   TimerOutputs v0.5.3
  [b77e0a4c]   InteractiveUtils
  [8f399da3]   Libdl
  [44cfe95a]   Pkg
  [de0858da]   Printf

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.0.1
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.2
  [be33ccc6]   CUDAnative v3.0.2
  [864edb3b] ↑ DataStructures v0.17.10 ⇒ v0.17.11
  [0c68f7d7]   GPUArrays v3.1.0 #cb79e08 (https://github.com/JuliaGPU/GPUArrays.jl.git)
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9]   TimerOutputs v0.5.3
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

kshyatt committed:
Allow scalar indexing where necessary and add a few tests

maleadt committed:
Merge pull request #674 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Merge pull request #621 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Update Project.toml

maleadt committed:
Move compiler functionality to GPUCompiler.jl.

maleadt committed:
Fix tests.

kshyatt committed:
Update src/execution.jl

Co-Authored-By: Tim Besard <tim.besard@gmail.com>

kshyatt committed:
Update src/memory.jl

Co-Authored-By: Tim Besard <tim.besard@gmail.com>

maleadt committed:
Move dynamic parallelism.

kshyatt committed:
Fix scalar indexing

kshyatt committed:
Move sentences

maleadt committed:
Adapt to GPUCompiler changes.

maleadt committed:
Wipe the runtime when packages change.

maleadt committed:
Move most of reflection to GPUCompiler.

maleadt committed:
Move the compilation cache to GPUCompiler.

kshyatt committed:
use the allowscalar macro

maleadt committed:
Pass FunctionSpecs to the compiler.

maleadt committed:
Adapt to minor output changes.

bors[bot] committed:
Merge #675

675: Allow scalar indexing where necessary and add a few tests r=maleadt a=kshyatt



Co-authored-by: Katharine Hyatt <khyatt@flatironinstitute.org>

kshyatt committed:
use a valid device for prefetch

maleadt committed:
Make sure array intrinsics propagate inboundness.

maleadt committed:
Allow NVTX.at-range with short function definitions.

maleadt committed:
Merge pull request #625 from JuliaGPU/tb/nvtx_short_functions

Allow NVTX.at-range with short function definitions.

maleadt committed:
Merge pull request #624 from JuliaGPU/tb/inbounds_array_intrinsics

Make sure array intrinsics propagate inboundness.

maleadt committed:
Bump version

maleadt committed:
Adapt to PTXCompiler changes.

maleadt committed:
Add back runtime build function for CI compatibility.

maleadt committed:
Move tests to GPUCompiler.

github-actions[bot] committed:
Update dependencies.

Project CUDAnative v3.0.4
    Status `~/work/CUDAnative.jl/CUDAnative.jl/Project.toml`
  [79e6a3ab]   Adapt v1.0.1
  [b99e7846]   BinaryProvider v0.5.8
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.2
  [f68482b8] ↑ Cthulhu v1.0.1 ⇒ v1.0.2
  [864edb3b] ↑ DataStructures v0.17.11 ⇒ v0.17.12
  [e2ba6199]   ExprTools v0.1.0
  [929cbde3]   LLVM v1.3.4
  [a759f4b9]   TimerOutputs v0.5.3
  [b77e0a4c]   InteractiveUtils
  [8f399da3]   Libdl
  [44cfe95a]   Pkg
  [de0858da]   Printf

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.0.1
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.2
  [be33ccc6] ↑ CUDAnative v3.0.2 ⇒ v3.0.4
  [864edb3b] ↑ DataStructures v0.17.11 ⇒ v0.17.12
  [0c68f7d7]   GPUArrays v3.1.0 #cb79e08 (https://github.com/JuliaGPU/GPUArrays.jl.git)
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9]   TimerOutputs v0.5.3
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #681 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Merge pull request #626 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Don't run problematic test on ARM CI.

bors[bot] committed:
Merge #604

604: Actually test error throwing for lsvqr r=maleadt a=kshyatt

Can't test mismatched dimensions if the GPU doesn't know about them

Co-authored-by: Katharine Hyatt <khyatt@flatironinstitute.org>

maleadt committed:
Merge pull request #231 from JuliaGPU/ksh/tests

Bunch of small tests and some doc fixes

matsueushi committed:
Wrapper functions for NNlib (#615)

mwarusz committed:
Propagate inbounds when using Const/ldg

maleadt committed:
Use the maximal supported capability with reflection utilities.

maleadt committed:
Simplify runtests.

maleadt committed:
Allow the debug build to fail, for now.

vchuravy committed:
Merge pull request #627 from mwarusz/mw/propagate-inbounds-ldg

Propagate inbounds when using Const/ldg

maleadt committed:
Build the asserts test with 1.4.0 where the BB LLVM asserts link works.

maleadt committed:
Merge pull request #623 from JuliaGPU/tb/gpucompiler

Move compiler functionality to GPUCompiler.jl

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.0.1
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.2
  [be33ccc6]   CUDAnative v3.0.4
  [864edb3b] ↑ DataStructures v0.17.12 ⇒ v0.17.13
  [0c68f7d7]   GPUArrays v3.1.0 #cb79e08 (https://github.com/JuliaGPU/GPUArrays.jl.git)
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9] ↑ TimerOutputs v0.5.3 ⇒ v0.5.4
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #686 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Bump version

AStupidBear committed:
fix 1d convolution

amontoison committed:
Fix sparse mul!

AStupidBear committed:
add comment to fix1d

maleadt committed:
Adapt to GPUCompiler changes.

maleadt committed:
Bump GPUCompiler.

maleadt committed:
Merge pull request #631 from JuliaGPU/tb/composition

Adapt to GPUCompiler changes.

amontoison committed:
Improve mul! testset names

maleadt committed:
Earlier stub library detection.

maleadt committed:
Small memory-related fixes.

maleadt committed:
Skip a sometimes broken test.

maleadt committed:
Merge pull request #233 from JuliaGPU/tb/mem_fixes

Small memory-related fixes.

maleadt committed:
Merge pull request #232 from JuliaGPU/tb/stub

Earlier stub library detection.

maleadt committed:
Disable the GC after taking pool-related spinlocks.

maleadt committed:
Merge pull request #690 from AStupidBear/conv1d

fix 1d convolution

maleadt committed:
Merge pull request #572 from janEbert/sparse

Implement a few sparse array basics

maleadt committed:
Use released GPUArrays.

maleadt committed:
Use Base.julia_cmd

maleadt committed:
Simplify use of Base.julia_cmd

maleadt committed:
Merge pull request #234 from JuliaGPU/tb/julia_cmd

Use Base.julia_cmd

maleadt committed:
Bump version

maleadt committed:
Merge pull request #634 from JuliaGPU/tb/julia_cmd

Simplify use of Base.julia_cmd

maleadt committed:
Only disable finalizers, not the entire GC.

maleadt committed:
Enable finalizers only after releasing the lock.

maleadt committed:
Update src/memory.jl

Co-Authored-By: Valentin Churavy <vchuravy@users.noreply.github.com>

amontoison committed:
Fix A' * x products for complex CuSparseMatrixCSC

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.1.0
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814] ↑ CUDAdrv v6.2.2 ⇒ v6.2.3
  [be33ccc6]   CUDAnative v3.0.4
  [864edb3b]   DataStructures v0.17.13
  [0c68f7d7]   GPUArrays v3.2.0
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9] ↑ TimerOutputs v0.5.4 ⇒ v0.5.5
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #694 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Move execution helpers to GPUCompiler.

maleadt committed:
Merge pull request #636 from JuliaGPU/tb/gpcompiler_execution

Move execution helpers to GPUCompiler.

bors[bot] committed:
Merge #692

692: Fix sparse mul! r=amontoison a=amontoison

close #629 
#630 , #637 
@haampie 

Co-authored-by: Alexis Montoison <alexis.montoison@polymtl.ca>

electronsandstuff committed:
adds wrappers for syevjBatched/heevjBatched family of CUSOLVER functions

maleadt committed:
Merge pull request #693 from JuliaGPU/tb/disable_gc

Disable the GC after taking pool-related spinlocks.

jw3126 committed:
add @cushow macro

maleadt committed:
Merge pull request #695 from electronsandstuff/electronsandstuff/syevjBatched

adds wrappers for syevjBatched/heevjBatched family of CUSOLVER functions

maleadt committed:
Wrap functionality to set and query stream priorities.

maleadt committed:
Don't require capability 7.5 for testing.

maleadt committed:
Don't require capability 7.5 for testing.

maleadt committed:
Move CuDim defs into separate file.

maleadt committed:
Wrap and test 3d memcpy.

maleadt committed:
Don't require capability 7.5 for testing.

maleadt committed:
Wrap and test host launch.

maleadt committed:
Merge pull request #637 from jw3126/cushow

add @cushow macro

maleadt committed:
Use Base.julia_cmd better.

maleadt committed:
Really don't require capability 7.5 for testing.

maleadt committed:
Merge pull request #639 from JuliaGPU/tb/update

Use Base.julia_cmd better.

github-actions[bot] committed:
Update dependencies.

Project CUDAnative v3.0.4
    Status `~/work/CUDAnative.jl/CUDAnative.jl/Project.toml`
  [79e6a3ab]   Adapt v1.0.1
  [b99e7846] ↑ BinaryProvider v0.5.8 ⇒ v0.5.9
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814] ↑ CUDAdrv v6.2.2 ⇒ v6.2.3
  [e2ba6199] ↑ ExprTools v0.1.0 ⇒ v0.1.1
  [61eb1bfa]   GPUCompiler v0.1.0 #6128874 (https://github.com/JuliaGPU/GPUCompiler.jl)
  [929cbde3]   LLVM v1.3.4
  [8f399da3]   Libdl
  [44cfe95a]   Pkg
  [de0858da]   Printf

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.1.0
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155]   CEnum v0.2.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.2.3
  [be33ccc6]   CUDAnative v3.0.4
  [864edb3b] ↑ DataStructures v0.17.13 ⇒ v0.17.15
  [0c68f7d7]   GPUArrays v3.2.0
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9]   TimerOutputs v0.5.5
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #702 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Repopulate the pool from freed blocks before allocating.

maleadt committed:
Add missing lock.

maleadt committed:
Merge pull request #704 from JuliaGPU/tb/binned_repopulate

Repopulate the pool from freed blocks before allocating.

maleadt committed:
Merge pull request #237 from JuliaGPU/tb/stream_priority

Wrap functionality to set and query stream priorities.

maleadt committed:
Merge pull request #239 from JuliaGPU/tb/launch_host

Wrap and test host launch.

maleadt committed:
Add GPUCompiler bounds.

maleadt committed:
Merge pull request #628 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Adapt to CEnum conversion changes.

https://github.com/JuliaInterop/CEnum.jl/pull/7

maleadt committed:
Merge pull request #240 from JuliaGPU/tb/convert_cenum

Adapt to CEnum conversion changes.

maleadt committed:
CEnum 0.3 compatibility.

maleadt committed:
CEnum 0.3 compatibility.

maleadt committed:
Merge pull request #705 from JuliaGPU/tb/cenum

CEnum 0.3 compatibility.

maleadt committed:
Merge pull request #640 from JuliaGPU/tb/cenum

CEnum 0.3 compatibility.

maleadt committed:
Avoid address space casts.

maleadt committed:
Improve API and add 3D test.

maleadt committed:
Only run WMMA tests on 1.4.1+

maleadt committed:
Make the source and desination positions 1-indexed.

maleadt committed:
Add a docstring.

aterenin committed:
Merge pull request #1 from JuliaGPU/master

Merge upstream changes.

maleadt committed:
Merge pull request #642 from JuliaGPU/tb/rm_addrspacecast

Avoid address space casts.

aterenin committed:
Add direct BLAS calls for trmm! and trsm!.

maleadt committed:
Add another test.

maleadt committed:
Merge pull request #238 from JuliaGPU/tb/memcpy_3d

Wrap 3d memcpy

maleadt committed:
Bump version

maleadt committed:
Protect against recursive failures during init.

maleadt committed:
Merge pull request #241 from JuliaGPU/tb/recursive_init

Protect against recursive failures during init.

maleadt committed:
Fixes for 1.5.

aterenin committed:
Add tests for BLAS.trmm!.

aterenin committed:
Add test for BLAS.trsm!.

aterenin committed:
Added mul! for trmm! and tests.

maleadt committed:
Merge pull request #643 from JuliaGPU/tb/1.5

Fixes for 1.5.

maleadt committed:
Merge pull request #707 from aterenin/master

Add direct BLAS calls for trmm! and trsm!

maleadt committed:
Merge pull request #708 from aterenin/master

Add mul! for trmm! and tests.

maleadt committed:
Adapt to GPUCompiler changes.

maleadt committed:
Support and use broadcast with mapreduce.

maleadt committed:
Protect against recursive failures during init.

maleadt committed:
Protect against recursive failures during init.

maleadt committed:
Merge pull request #645 from JuliaGPU/tb/recursive_init

Protect against recursive failures during init.

maleadt committed:
Merge pull request #710 from JuliaGPU/tb/recursive_init

Protect against recursive failures during init.

maleadt committed:
Merge pull request #709 from JuliaGPU/tb/mapreduce_broadcast

Support and use broadcast with mapreduce.

maleadt committed:
Perform threaded test in lock.

maleadt committed:
Merge pull request #644 from JuliaGPU/tb/redesign_gpucompilers

Adapt to GPUCompiler changes.

maleadt committed:
Merge pull request #711 from JuliaGPU/tb/threaded_testset

Perform threaded test in lock.

maleadt committed:
Bump GPUCompiler.

bsmiles32 committed:
findfirst should return nothing if no matching element is found

maleadt committed:
Bump version

maleadt committed:
Update Manifest, use released GPUArrays.

maleadt committed:
Bump version

maleadt committed:
Merge pull request #712 from bsmiles32/fix/findfirst

findfirst should return nothing if no matching element is found

github-actions[bot] committed:
Update dependencies.

Project CUDAnative v3.1.0
    Status `~/work/CUDAnative.jl/CUDAnative.jl/Project.toml`
  [79e6a3ab]   Adapt v1.0.1
  [b99e7846]   BinaryProvider v0.5.9
  [fa961155] ↑ CEnum v0.2.0 ⇒ v0.3.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814]   CUDAdrv v6.3.0
  [e2ba6199]   ExprTools v0.1.1
  [61eb1bfa]   GPUCompiler v0.2.0
  [929cbde3]   LLVM v1.4.1
  [8f399da3]   Libdl
  [44cfe95a]   Pkg
  [de0858da]   Printf

github-actions[bot] committed:
Update dependencies.

Project CuArrays v2.2.0
    Status `~/work/CuArrays.jl/CuArrays.jl/Project.toml`
  [621f4979]   AbstractFFTs v0.5.0
  [79e6a3ab]   Adapt v1.0.1
  [fa961155] ↑ CEnum v0.2.0 ⇒ v0.3.0
  [3895d2a7]   CUDAapi v4.0.0
  [c5f51814] ↑ CUDAdrv v6.2.3 ⇒ v6.3.0
  [be33ccc6] ↑ CUDAnative v3.0.4 ⇒ v3.1.0
  [864edb3b]   DataStructures v0.17.15
  [0c68f7d7]   GPUArrays v3.3.0
  [1914dd2f]   MacroTools v0.5.5
  [872c559c]   NNlib v0.6.6
  [189a3867]   Reexport v0.2.0
  [ae029012]   Requires v1.0.1
  [a759f4b9]   TimerOutputs v0.5.5
  [8f399da3]   Libdl
  [37e2e46d]   LinearAlgebra
  [44cfe95a]   Pkg
  [de0858da]   Printf
  [9a3f8284]   Random
  [2f01184e]   SparseArrays
  [10745b16]   Statistics

maleadt committed:
Merge pull request #715 from JuliaGPU/update_manifest

Update manifest

maleadt committed:
Merge pull request #647 from JuliaGPU/update_manifest

Update manifest

marius311 committed:
fix ambiguity for Random.seed!(rng, nothing)

maleadt committed:
Actually put the workspace buffer back in the memory pool.

maleadt committed:
Merge pull request #717 from marius311/seedambiguity

fix ambiguity for Random.seed!(rng, nothing)

maleadt committed:
Remove broken, unused, DNN wrapper.

maleadt committed:
Merge pull request #718 from JuliaGPU/tb/unsafe_free_workspace

Actually put the workspace buffer back in the memory pool.

github-actions[bot] committed:
CompatHelper: bump compat for "CEnum" to "0.4"

github-actions[bot] committed:
CompatHelper: bump compat for "CEnum" to "0.4"

maleadt committed:
Merge pull request #720 from JuliaGPU/compathelper/new_version/2020-05-15-00-09-48-579-720983604

CompatHelper: bump compat for "CEnum" to "0.4"

maleadt committed:
Merge pull request #650 from JuliaGPU/compathelper/new_version/2020-05-15-00-10-36-611-306955163

CompatHelper: bump compat for "CEnum" to "0.4"

maleadt committed:
Allow nightly to fail.

maleadt committed:
Remove debug statements.

maleadt committed:
Fix return type of report_oom.

Ref JuliaGPU/GPUCompiler.jl#32

maleadt committed:
Add the exception flag later.

maleadt committed:
Simplify test using CuHostArray.

maleadt committed:
Add 'CUDAdrv.jl/' from commit '0c1b23b3725750ce7d521b014cec9089f637c685'

git-subtree-dir: CUDAdrv.jl
git-subtree-mainline: 72f7ec3e4eb6e94e02ddfcd400b800d3a32b19ab
git-subtree-split: 0c1b23b3725750ce7d521b014cec9089f637c685

maleadt committed:
Add 'CUDAapi.jl/' from commit 'fcf574076ddfba6f5689c96ec4d00bc6ec51ccff'

git-subtree-dir: CUDAapi.jl
git-subtree-mainline: ac45499a9a0f886c47a01bd23db6b7796458b408
git-subtree-split: fcf574076ddfba6f5689c96ec4d00bc6ec51ccff

maleadt committed:
Add 'CUDAnative.jl/' from commit 'f90eb4c5b81dc1149997ffa8ee609daa965a7fdb'

git-subtree-dir: CUDAnative.jl
git-subtree-mainline: 0401a662569bfc6190744db45c1047d4bbf2f2a0
git-subtree-split: f90eb4c5b81dc1149997ffa8ee609daa965a7fdb

maleadt committed:
Add 'CuArrays.jl/' from commit 'fc712f0ca369c998029543ca4e46e9b5fa0a509e'

git-subtree-dir: CuArrays.jl
git-subtree-mainline: c0eca88d5170ddc59de728b2b0e8cf6bc0190813
git-subtree-split: fc712f0ca369c998029543ca4e46e9b5fa0a509e

maleadt committed:
Reorganize into a single package.

maleadt committed:
Run tests in parallel.

maleadt committed:
Remove broken test of linear memcpy3d.

maleadt committed:
Add CUDA version of Base.at-timed, and use it to report GC stats during testing.

maleadt committed:
Remove reverse testing of Flux.

maleadt committed:
Test and silence warning message.

maleadt committed:
Remove threaded timings workaround.

maleadt committed:
Split and add some tests.

maleadt committed:
Iterate the GPUArrays tests instead of hard-coding them.

maleadt commented:
Hmm, this failure seems to be happening a lot:

```
Error During Test at /builds/JuliaGPU/CUDA.jl/test/array.jl:407
  Got exception outside of a @test
  TaskFailedException:
  CUDA error: an illegal memory access was encountered (code 700, ERROR_ILLEGAL_ADDRESS)
  Stacktrace:
   [1] throw_api_error(::CUDA.cudaError_enum) at /builds/JuliaGPU/CUDA.jl/lib/cuda/error.jl:103
   [2] macro expansion at /builds/JuliaGPU/CUDA.jl/lib/cuda/error.jl:110 [inlined]
   [3] cuMemcpyDtoH_v2(::Ptr{Float32}, ::CuPtr{Float32}, ::Int64) at /builds/JuliaGPU/CUDA.jl/lib/utils/call.jl:93
   [4] #unsafe_copyto!#6 at /builds/JuliaGPU/CUDA.jl/lib/cuda/memory.jl:324 [inlined]
   [5] unsafe_copyto! at /builds/JuliaGPU/CUDA.jl/lib/cuda/memory.jl:317 [inlined]
   [6] unsafe_copyto! at /builds/JuliaGPU/CUDA.jl/src/array.jl:309 [inlined]
   [7] copyto!(::Array{Float32,2}, ::Int64, ::CuArray{Float32,2,Nothing}, ::Int64, ::Int64) at /builds/JuliaGPU/CUDA.jl/src/array.jl:284
   [8] copyto! at /builds/JuliaGPU/CUDA.jl/.julia/packages/GPUArrays/OXvxB/src/host/abstractarray.jl:112 [inlined]
   [9] Array at ./array.jl:541 [inlined]
   [10] Array(::CuArray{Float32,2,Nothing}) at ./boot.jl:429
   [11] macro expansion at /builds/JuliaGPU/CUDA.jl/test/array.jl:421 [inlined]
   [12] (::var"#1588#threadsfor_fun#169"{ReentrantLock,UnitRange{Int64}})(::Bool) at ./threadingconstructs.jl:61
   [13] (::var"#1588#threadsfor_fun#169"{ReentrantLock,UnitRange{Int64}})() at ./threadingconstructs.jl:28
```

Even though this PR didn't really touch any of that code. Maybe because of the distributed execution?

maleadt committed:
Streamline adding workers.

maleadt committed:
Import CUPTI for CUPTIError.

maleadt committed:
Support for running under cuda-memcheck.

maleadt committed:
Fix testset naming.

maleadt committed:
Skip some tests when running under cuda-memcheck.

maleadt committed:
Fix passing TestSuite closures under 1.5

maleadt committed:
Print start of testsuite when running under cuda-memcheck.

maleadt committed:
Use arrays instead of raw pointers for some tests.

This also fixes an OOB.

maleadt committed:
Simplify some more tests.

maleadt committed:
Disable threading test.

maleadt committed:
Update the documentation.

maleadt committed:
Disable failing tests.

maleadt committed:
Add README. [ci skip]
